### PR TITLE
Convert DanglingOnly to Filters for `docker image prune`

### LINKED ISSUE #28497: Prune based on age
https://github.com/docker/docker/pull/26108 is fantastic. However, it doesn't handle this use case: a CI server where I need to keep images around for a couple days for caching, but need to prune them after that. I would like to have a time parameter (in days or hours) to exclude image/container/etc pruning. An LRU system would be ideal, but more technically complicated. For context, I am handling this with the script below. It's a little gross, but gets the job done.
```
#!/bin/bash
function check_age(){
    while read cid
    do
        age=$(date -d $(docker inspect -f '{{.Created}}' $cid) +%s)
        if [ $age -lt $TOO_OLD ]; then
            echo "$cid"
        fi
    done
}

get_names(){
    while read cid
    do
        docker inspect -f '{{.RepoTags}}' $cid | tr -d '[]'
    done
}

echo "cleaning containers"
TOO_OLD=$(echo "$(date +%s)-60*60*6" | bc)
docker rm -v $(docker ps --filter status=exited -q 2>/dev/null)
docker rm -f $(docker ps -aq --no-trunc | check_age)

echo "cleaning images"
docker rmi $(docker images --no-trunc -q --filter "dangling=true")
TOO_OLD=$(echo "$(date +%s)-60*60*24*7" | bc)
docker rmi $(docker images -q --no-trunc | check_age | get_names)

exit 0
```

### PR DESCRIPTION
**Pull request by yongtang**
**- What I did**

This fix is part of fix for #28497

This fix convert DanglingOnly in ImagesPruneConfig to Filters, so that it is possible to maintain API compatibility in the future.

A follow up to this PR will be done once this PR is merged.

**- How I did it**

**- How to verify it**

Several integration tests have been added to cover changes.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

cute-baby-animal-wallpapers-hd-images

This fix is related to #28497.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
SGTM

It would be great if we can support --until="5min ago" as well.

Also it would be great if we can have support for network (another PR)

**Comment by thaJeztah** [comment]
Should we use `--filter` for this? I think it's more consistent, and expandable (`docker prune --filter label=foo` at some point?)

**Comment by AkihiroSuda** [comment]
IMO we should have both filter and until for CLI, but for API, we only need filter.

**Comment by vdemeester** [comment]
I do agree with @thaJeztah, it's the same as `images` where we had `--until` and `--since` and we unified it under `--filter`

**Comment by justincormack** [comment]
SGTM, current prune has not been very useful for CI machines so far, really want prune by age for that use case.

**Comment by mlaventure** [comment]
I'm with @vdemeester and @thaJeztah can we have a `--filter` instead?

**Comment by yongtang** [comment]
Thanks all. Let me update the PR based on the comments.

**Comment by yongtang** [comment]
I noticed that currently, `images/prune` has a `DanglingOnly` field:

```
// POST "/images/prune"
type ImagesPruneConfig struct {
  	DanglingOnly bool	
  }
```

However, if we takes a filter, then this field could very well be replaced with `dangling=true`, which will be consistent with `docker volume ls --filter dangling=true`.

Should we convert `DanglingOnly` to` filter` as well?

**Comment by mlaventure** [comment]
@yongtang sure, why not.

**Comment by thaJeztah** [comment]
Would `dangling=false` work though (for `prune`)?

**Comment by mlaventure** [comment]
@thaJeztah both should work.

Currently danglingOnly default to true and `-a` switch it. It also only affects images atm.

**Comment by yongtang** [comment]
The PR has been updated. Now the `--filter` flag takes both `dangling=<true|false>` and `until=<timestamp>`, replacing `DanglingOnly bool`. That could be an API change. Depending on if this PR is considered for 1.13 or later, additional update might be required to maintain API compatibility.

File: cli/command/container/prune.go [hunk 1]
```diff
@@ -40,6 +42,7 @@ func NewPruneCommand(dockerCli *command.DockerCli) *cobra.Command {
 
 	flags := cmd.Flags()
 	flags.BoolVarP(&opts.force, "force", "f", false, "Do not prompt for confirmation")
+	flags.Var(&opts.filter, "filter", "Provide filter values (i.e. 'until=<timestamp>')")
```
Comment (AkihiroSuda) [inline cli/command/container/prune.go]: i.e. -> e.g.

File: cli/command/system/prune.go [hunk 2]
```diff
@@ -49,6 +52,28 @@ Are you sure you want to continue?`
 )
 
 func runPrune(dockerCli *command.DockerCli, opts pruneOptions) error {
+	// --all and --filter all=<true|false> should match
```
Comment (AkihiroSuda) [inline cli/command/system/prune.go]: Let's dedup this "all vs dangling" logic

File: cli/command/system/prune.go [hunk 3]
```diff
@@ -32,6 +34,7 @@ func NewPruneCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags := cmd.Flags()
 	flags.BoolVarP(&opts.force, "force", "f", false, "Do not prompt for confirmation")
 	flags.BoolVarP(&opts.all, "all", "a", false, "Remove all unused images not just dangling ones")
+	flags.Var(&opts.filter, "filter", "Provide filter values (i.e. 'until=<timestamp>')")
```
Comment (AkihiroSuda) [inline cli/command/system/prune.go]: i.e. -> e.g.

File: cli/command/system/prune.go [hunk 4]
```diff
@@ -63,8 +88,16 @@ func runPrune(dockerCli *command.DockerCli, opts pruneOptions) error {
...
+		return err
+	}
+	spaceReclaimed += spc
+	if output != "" {
+		fmt.Fprintln(dockerCli.Out(), output)
+	}
+
 	for _, pruneFn := range []func(dockerCli *command.DockerCli) (uint64, string, error){
-		prune.RunContainerPrune,
 		prune.RunVolumePrune,
```
Comment (AkihiroSuda) [inline cli/command/system/prune.go]: can you support filter for volume and network as well?
(can be another PR)

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]

File: cli/command/image/prune.go [hunk 5]
```diff
@@ -8,18 +8,20 @@ import (
...
+	"github.com/docker/docker/opts"
 	units "github.com/docker/go-units"
 	"github.com/spf13/cobra"
 )
 
 type pruneOptions struct {
-	force bool
-	all   bool
+	force  bool
+	all    bool // TODO: deprecate and in favor of --filter all=true
```
Comment (AkihiroSuda) [inline cli/command/image/prune.go]: I think we don't need to deprecate `--all`, because is short and easy to use

**Comment by AkihiroSuda** [comment]
> That could be an API change. Depending on if this PR is considered for 1.13 or later, additional update might be required to maintain API compatibility.

Since 1.13 has been already freezed, I do think the `until=<timestamp>` should not be in 1.13.
However, I also understand the issue about maintaining compatibility of the API.
So, how about the following plan?

### 1.13
- API:  Replace `ImagePruneConfig.DanglingOnly` with `ImagePruneConfig.Filters` right now. However, the filter API should only support `dangling=true/false` at the moment, and should not support `until=<timestamp>`.
- CLI: Do not add `--filter` at the moment. However, the `--all` option should be internally converted to `ImagePruneConfig.Filters`.

### 1.14
- API: Add `until=<timestamp>` (and so on, if we want)
- CLI: Add `--filter`.

WDYT?

**Comment by mlaventure** [comment]
I know we've frozen the branch yet, but I think i'd rather we make an exception and replace `DanglingOnly` with `Filter` to avoid having to maintain both in the future since this API wasn't in any stable release yet.

**Comment by thaJeztah** [comment]
Agree with @AkihiroSuda and @mlaventure (discussed briefly with @AkihiroSuda earlier today). It's not _released_ yet, so we can still modify. It's better done now, than having to officially deprecate / change after release

**Comment by yongtang** [comment]
Thanks all for the review. I am also in favor of replacing `DanglingOnly` with `Filter` and get it in before 1.13 release (just to avoid future deprecate/compatibility issue).

I will update the PR based on the comments shortly.

**Comment by yongtang** [comment]
Thanks all for the review. The PR has been updated and the scope of this PR has been reduced to only convert `DanglingOnly` to `Filters`. This is to align with the discussion so that it is possible to get this one in 1.13 and maintain API compatibility in the future.

One this PR is merged I will open another PR to add the `until` filter (possibly for post-1.13/1.14)

Please take a look and let me know if there are any issues.

**Comment by thaJeztah** [comment]
Thanks @yongtang! Can you update the PR's title and description?

File: daemon/prune.go [hunk 6]
```diff
@@ -73,8 +74,17 @@ func (daemon *Daemon) VolumesPrune(config *types.VolumesPruneConfig) (*types.Vol
 func (daemon *Daemon) ImagesPrune(config *types.ImagesPruneConfig) (*types.ImagesPruneReport, error) {
 	rep := &types.ImagesPruneReport{}
 
+	danglingOnly := true
+	if config.Filters.Include("dangling") {
+		if config.Filters.ExactMatch("dangling", "false") || config.Filters.ExactMatch("dangling", "0") {
+			danglingOnly = false
+		} else if !(config.Filters.ExactMatch("dangling", "true") || config.Filters.ExactMatch("dangling", "1")) {
+			return nil, fmt.Errorf("Flag --filter dangling= should only be true or false")
```
Comment (AkihiroSuda) [inline daemon/prune.go]: nit: you can use errors.New 

Others LGTM, thank you!

**Review by AkihiroSuda — APPROVED** [review]

File: daemon/prune.go (hunk 6 above)
Comment (yongtang) [inline daemon/prune.go]: Thanks @AkihiroSuda the PR has been updated.

**Comment by yongtang** [comment]
Thanks @thaJeztah. The PR's title and description has been updated.

File: api/types/types.go [hunk 7]
```diff
@@ -512,12 +512,13 @@ type DiskUsage struct {
 // ImagesPruneConfig contains the configuration for Engine API:
 // POST "/images/prune"
 type ImagesPruneConfig struct {
-	DanglingOnly bool
+	Filters filters.Args
 }
 
 // ContainersPruneConfig contains the configuration for Engine API:
 // POST "/images/prune"
```
Comment (thaJeztah) [inline api/types/types.go]: looks like the wrong endpoint is mentioned here

File: api/types/types.go [hunk 8]
```diff
@@ -512,12 +512,13 @@ type DiskUsage struct {
...
 type ContainersPruneConfig struct {
+	Filters filters.Args
 }
 
 // VolumesPruneConfig contains the configuration for Engine API:
```
Comment (thaJeztah) [inline api/types/types.go]: The line below this as well (wrong endpoint)

File: api/types/types.go [hunk 9]
```diff
@@ -512,12 +512,13 @@ type DiskUsage struct {
 // ImagesPruneConfig contains the configuration for Engine API:
 // POST "/images/prune"
 type ImagesPruneConfig struct {
-	DanglingOnly bool
+	Filters filters.Args
```
Comment (thaJeztah) [inline api/types/types.go]: Don't forget to update the swagger.yml do match this change; https://github.com/docker/docker/blob/22ba220e864d6d30ce1ed56a5461b96dbade844f/api/swagger.yaml#L4830-L4868

File: api/server/router/container/container_routes.go [hunk 10]
```diff
@@ -545,16 +545,16 @@ func (s *containerRouter) postContainersPrune(ctx context.Context, w http.Respon
...
-	if err := httputils.CheckForJSON(r); err != nil {
+	pruneFilters, err := filters.FromParam(r.Form.Get("filters"))
+	if err != nil {
 		return err
 	}
 
-	var cfg types.ContainersPruneConfig
-	if err := json.NewDecoder(r.Body).Decode(&cfg); err != nil {
-		return err
+	cfg := &types.ContainersPruneConfig{
```
Comment (thaJeztah) [inline api/server/router/container/container_routes.go]: Should these changes for containers be on the other (follow up) PR?

File: api/types/types.go [hunk 11]
```diff
@@ -512,12 +512,13 @@ type DiskUsage struct {
...
 }
 
 // ContainersPruneConfig contains the configuration for Engine API:
 // POST "/images/prune"
 type ContainersPruneConfig struct {
+	Filters filters.Args
```
Comment (thaJeztah) [inline api/types/types.go]: Same here

File: client/container_prune.go [hunk 12]
```diff
@@ -3,8 +3,10 @@ package client
 import (
 	"encoding/json"
 	"fmt"
+	"net/url"
```
Comment (thaJeztah) [inline client/container_prune.go]: Changes here are probably for the other PR as well?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Thanks @yongtang, I left some notes/questions

**Comment by yongtang** [comment]
Thanks @thaJeztah. I updated the PR to address the endpoint mismatch issue (e.g, `POST "/images/prune"` vs `ContainersPruneConfig`)

Before I tried to update the swagger API, I noticed that,
- With `Filters` flag, we probably could cover any parameter for `ContainersPruneConfig/ImagesPruneConfig/VolumesPruneConfig/NetworksPruneConfig`.
- The `Filters` flag has to be passed as the query.

As a result, if we think we will use `Filters filters.Args` for any future parameters/extensions, then there is no need to use `POST` with a body. We only need to use `POST` with a empty body (and `Filters filters.Args` will be in query string).

I am wondering if we could simply drop the `...PruneConfig` and only pass a `filters.Args`? For example:
```
func (cli *Client) ContainersPrune(ctx context.Context, pruneFilter filters.Args) (types.ContainersPruneReport, error)
```
And then inside the func, we could:
```
        // only query based on filter, no `BODY` part in `POST`
	serverResp, err := cli.post(ctx, "/containers/prune", query, nil, nil) 
	if err != nil {
		return report, err
	}
	defer ensureReaderClosed(serverResp)
```

As long as we always use filter for any parameter passing in prune, then I think we should be fine.

**Comment by mlaventure** [comment]
Moving it all to the query SGTM since I couldn't think of options that wouldn't fit the filter category for prune.

**Comment by thaJeztah** [comment]
Let me move this to "must have" for 1.13, otherwise we have to go through a deprecation cycle / API change

**Comment by yongtang** [comment]
Thanks all for the review. The PR has been updated with Filters replacing Config types. The swagger API has also been updated. Please take a look and let me know if there are any issues.

**Comment by thaJeztah** [comment]
@yongtang looks like there's an issue with swagger;

```
19:23:04 The swagger spec at "api/swagger.yaml" is valid against swagger specification 2.0
19:23:06 The result of hack/generate-swagger-api.sh differs
19:23:06 
19:23:06 ?? api/types/containers/
19:23:06 
19:23:06 Please update api/swagger.yaml with any api changes, then 
19:23:06 run `hack/generate-swagger-api.sh`
```

**Comment by yongtang** [comment]
The swagger seems to be related to #29045. Not sure why it always generating a `api/types/containers`.

File: client/image_prune.go [hunk 13]
```diff
@@ -3,20 +3,31 @@ package client
...
-func (cli *Client) ImagesPrune(ctx context.Context, cfg types.ImagesPruneConfig) (types.ImagesPruneReport, error) {
+func (cli *Client) ImagesPrune(ctx context.Context, pruneFilters filters.Args) (types.ImagesPruneReport, error) {
 	var report types.ImagesPruneReport
 
 	if err := cli.NewVersionError("1.25", "image prune"); err != nil {
 		return report, err
 	}
 
-	serverResp, err := cli.post(ctx, "/images/prune", nil, cfg, nil)
+	query := url.Values{}
```
Comment (mlaventure) [inline client/image_prune.go]: This could be moved in `utils.go` since it's duplicated code.

Also, for consistency, could this also be added to `client.VolumesPrune()` and `client.NetworksPrune()` even though they don't have any yet?

**Review by mlaventure — CHANGES_REQUESTED** [review]

**Comment by yongtang** [comment]
Thanks @mlaventure for the review. The PR has been updated with duplicated code moved into utils.go. client.VolumesPrune() and client.NetworksPrune() have also been updated. Please take a look and let me know if there are any issues.

File: daemon/prune.go [hunk 14]
```diff
@@ -70,11 +72,20 @@ func (daemon *Daemon) VolumesPrune(config *types.VolumesPruneConfig) (*types.Vol
...
-func (daemon *Daemon) ImagesPrune(config *types.ImagesPruneConfig) (*types.ImagesPruneReport, error) {
+func (daemon *Daemon) ImagesPrune(pruneFilters filters.Args) (*types.ImagesPruneReport, error) {
 	rep := &types.ImagesPruneReport{}
 
+	danglingOnly := true
+	if pruneFilters.Include("dangling") {
+		if pruneFilters.ExactMatch("dangling", "false") || pruneFilters.ExactMatch("dangling", "0") {
+			danglingOnly = false
+		} else if !(pruneFilters.ExactMatch("dangling", "true") || pruneFilters.ExactMatch("dangling", "1")) {
+			return nil, errors.New("Flag --filter dangling= should only be true or false")
```
Comment (cpuguy83) [inline daemon/prune.go]: Says it must be true or false here, but above we accept `0` for false?
Is there precedent in other filters for this?
Most boolean params we accept true/false/1/0

**Review by mlaventure — APPROVED** [review]

**Review by cpuguy83 — APPROVED** [review]
LGTM

ping @thaJeztah

**Review by vdemeester — APPROVED** [review]
LGTM :frog:

**Comment by nicosingh** [comment]
Hi,

I'm very very sorry in advance for this newbie question 😬, but I can't understand the definition of `dangling` parameter in `/images/prune` endpoint ([ref](https://docs.docker.com/engine/api/v1.40/#operation/ImagePrune)):

```
dangling=<boolean>:
- When set to true (or 1), prune only unused and untagged images
- When set to false (or 0), all unused images are pruned
```

The usage of the words **_only_** and **_all_** brings my confusion. I'd expect to read something like this (if my understanding is correct):

```
dangling=<boolean>:
- When set to true (or 1), both unused and untagged images are pruned
- When set to false (or 0), only unused images are pruned
```

Or:

```
dangling=<boolean>:
- When set to true (or 1), only untagged images are pruned
- When set to false (or 0), only unused images are pruned
```

Or:

```
dangling=<boolean>:
- When set to true (or 1), only untagged images are pruned
- When set to false (or 0), both unused and untagged images are pruned
```

By the way, which is the default value for `dangling` field?. I'd expect to think it's `false`.

Thanks in advance for your help! My apologies if these questions are duplicated (I couldn't find them elsewhere).
